library/java/org/chromium/net/impl/CancelProofEnvoyStream.java

```diff
@@ -0,0 +1,114 @@
+ * is invoked while the stream is being created, that cancel will be executed only once the stream
+ * is completely initialized. Doing otherwise leads to unpredictable outcomes.
+ */
+final class CancelProofEnvoyStream {
+
+  private static final int CANCEL_BIT = 0x8000;
+  private final AtomicInteger mState = new AtomicInteger(1); // Busy starting.
+  private volatile EnvoyHTTPStream mStream;                  // Cancel can come from any Thread.
+
+  /** Sets the stream. Can only be invoked once. */
+  void setStream(EnvoyHTTPStream stream) {
+    assert mStream == null;
+    mStream = stream;
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /** Initiates the sending of the request headers if the state permits. */
+  void sendHeaders(Map<String, List<String>> envoyRequestHeaders, boolean endStream) {
+    if (cancelHasAlreadyBeenInvoked()) {
+      return; // Already Cancelled - to late to send something.
+    }
+    mStream.sendHeaders(envoyRequestHeaders, endStream);
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /** Initiates the sending of one chunk of the request body if the state permits. */
+  void sendData(ByteBuffer buffer, boolean finalChunk) {
+    if (cancelHasAlreadyBeenInvoked()) {
+      return; // Already Cancelled - to late to send something.
+    }
+    // The Envoy Mobile library only cares about the capacity - must use the correct ByteBuffer
+    if (buffer.position() == 0) {
+      mStream.sendData(buffer, buffer.remaining(), finalChunk);
+    } else {
+      ByteBuffer resizedBuffer = ByteBuffer.allocateDirect(buffer.remaining());
+      buffer.mark();
+      resizedBuffer.put(buffer);
+      buffer.reset();
+      mStream.sendData(resizedBuffer, finalChunk);
+    }
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /** Initiates the reading of one chunk of the the request body if the state permits. */
+  void readData(int size) {
+    if (cancelHasAlreadyBeenInvoked()) {
+      return; // Already Cancelled - to late to read something.
+    }
+    mStream.readData(size);
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /**
+   * Cancels the Stream if the state permits. Will be delayed when an operation is concurrently
+   * running. Idempotent and Thread Safe.
+   */
+  void cancel() {
+    while (true) {
+      int count = mState.get();
+      if ((count & CANCEL_BIT) != 0) {
+        return; // Cancel already invoked.
+      }
+      if (mState.compareAndSet(count, count | CANCEL_BIT)) {
+        if (count == 0) {
+          mStream.cancel(); // Was not busy with other EM operations - cancel right now.
```

Comment:
That's CAS magic: https://en.wikipedia.org/wiki/Compare-and-swap

This technique is extensively used in Operating Systems and low level telecom logic. The whole PR is based on this.

The contract is the mutation succeeds only if the original value matches the expected one - this is atomic at the assembly language level: most CPUs have dedicated mnemonics for this operation - extremely efficient.

And this might look like an infinite loop. It is infinite only if many Threads are **eternally** attempting to concurrently change the value. In fact, CAS is pretty bad under heavy contention - in that case it is probably better to go with "synchronized" blocks. In our case, there is none or very little contention. What matters is correctness.

This PR requires CAS knowledge. I am not too sure how disseminate that knowledge.

Reply:
Yeah, I google'd "compare and set" and I think I understand that primitive (after swapping back in some CS knowledge from years ago :>) and how it is used in this PR. 

But CAS is merely a primitive. The reason it's OK to call cancel() here is because once the CANCEL_BIT is set, mState in unable to be incremented. It can only be decremented. So once mState = CANCEL_BIT it will never change. (Because that's the contract that the other methods agree to). Is that assessment correct?

As for infinite looping... This method will busy loop until any in-flight operations finish, right? As such if any of those operations need access to the current thread, they will deadlock, I think? Presumably, something prevents that from happening. Can you expand on what that is? Also it looks like setStream() finished an operation that gets started elsewhere. Is it possible that cancel() is called before setStream() but for some reason setStream() is blocked on the same thread that cancel() is called from? I think you've thought about this enough that this is all probably true, but it would be good to have this all written down.

Reply:
Discussed in person, and comments were added.